### ISSUE TITLE
`tutorial/rnn/translate` is not compatible with `r0.12`

### ISSUE THREAD

**Issue by bxshi** [tensorflow/models#846]
Current found problems:

1. Referencing deprecated `contrib` package
2. Referencing deprecated `seq2seq_legacy`
3. Error when calling `sampled_softmax_loss` in `seq2seq_model.py`
4. No gradient defined for `LogUniformCandidateSampler`

I'm working on a PR to fix this and please let me know if you found other problems.

**Comment by tjrileywisc** [comment]
Are you running into a problem with this line (line 122 in seq2seq_model.py)? That's where I get stopped:

single_cell = tf.contrib.rnn.GRUCell(size)

**Comment by bxshi** [comment]
This could be solved by change this line into 

`single_cell = tf.nn.rnn_cell.GRUCell`

You need to change other `contrib`s accordingly because the newest version of TF has removed this package.

Baoxu (Dash) Shi
Computer Science and Engineering Department
University of Notre Dame

https://www.linkedin.com/in/baoxushi

**Comment by tjrileywisc** [comment]
I see what you mean, thanks. I fixed those and got stopped somewhere else
(calling a function from legacy_seq2seq), so I guess I have some work to do
to find everything again.

**Comment by bxshi** [comment]
@tjrileywisc After that you may be stuck with some other errors about the rank of the tensors does not match. I'm trying to reproduce this problem and see if I can solve it. Please let me know if you know how to fix it. Thanks!

**Comment by tjrileywisc** [comment]
Are you seeing this? Because I think I've hit your problem as well. I had
to fiddle around with imports in __init__.py files and am no longer using
the legacy_seq2seq module:

File
"c:\python35\lib\site-packages\tensorflow\python\framework\common_shapes.py",
line 675, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Shape must be rank 2 but is rank 1 for
'model_with_buckets/sequence_loss/sequence_loss_by_example/sampled_softmax_loss/MatMul_1'
(op: 'MatMul') with input shapes: [?], [?,1024].

**Comment by bxshi** [comment]
@tjrileywisc Yes I have reproduced this, the current fix I found is 

update L109 in `seq2seq_model.py` to`local_inputs = tf.reshape(tf.cast(inputs, tf.float32), [-1, size])`.

The called function `sampled_softmax_loss` takes inputs in shape [batch_size, dim], but local_inputs was in [?], not [?, 1024].

However this gives me another error which is 

```
Traceback (most recent call last):
  File "/data/bshi/py3env/lib/python3.4/site-packages/tensorflow/python/ops/gradients_impl.py", line 455, in gradients
    grad_fn = ops.get_gradient_function(op)
  File "/data/bshi/py3env/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 1543, in get_gradient_function
    return _gradient_registry.lookup(op_type)
  File "/data/bshi/py3env/lib/python3.4/site-packages/tensorflow/python/framework/registry.py", line 93, in lookup
    "%s registry has no entry for: %s" % (self._name, name))
LookupError: gradient registry has no entry for: LogUniformCandidateSampler
```

I'm gonna try add a dummy gradient and see if that could be the fix. Any thoughts?

**Comment by bxshi** [comment]
[PR3414](https://github.com/tensorflow/tensorflow/pull/3414) is related to solve the `LogUniformCandidateSampler` gradient problem.

**Comment by tjrileywisc** [comment]
Okay, I added this to the bottom of candidate_sampling_ops.py (it seems
like that file has changed a bit since that PR):

from tensorflow.python.framework import ops
import tensorflow as tf
@ops.RegisterGradient("LogUniformCandidateSampler")
def _LogUniformCandidateSamplerGrad(op, *grads):
    # input: true_classes grad
    return [tf.cast(tf.zeros_like(grads[1]), tf.int64)]

It's parsing the data, we'll see if the model trains.

Thanks for your help on this!

**Comment by bxshi** [comment]
@tjrileywisc NP man! I'll try work out a PR soon so others don't need to go through all these!

Happy coding/researching!

**Comment by bxshi** [comment]
The current model works with tensorflow/tensorflow's master branch, so I'll close this issue.

**Comment by k0105** [comment]
Could you elaborate on that? I'm on Ubuntu 16.04 LTS and have both TensorFlow 0.12.1 (Python 2, installed via apt) and TensorFlow 1.0.0a4 (installed via pip and .whl file), but I still get the gradient error after having fixed all other ones. What can I do to properly run the seq2seq code?

Thanks in advance.

**Comment by madspark** [comment]
Use the latest nightly build or build from source. The latest master branch version seems to work fine.

**Comment by k0105** [comment]
Final report:
Thank you. I was confused that compiling from source only produces a .whl file that indicates version 0.12.1, which I had already installed, but when I installed this file in a new virtual environment it actually worked unlike my original 0.12.1 from pip. So if anyone runs into the same issue: Literally just clone the tensorflow repo Master branch, build from source and use the resulting .whl file without getting distracted by its name. [To be exact it now runs out of memory on a 1060 with 6GB, but I'll rerun it on a TITAN X cluster later tonight - 12x 12GB will hopefully suffice.]
